src/python/pants/engine/fs_test.py

```diff
@@ -443,6 +444,118 @@ def test_digest_entries_handles_empty_directory(rule_runner: RuleRunner) -> None
+        [
+            SymlinkEntry("a.ln", "a.txt"),
+            FileEntry(
+                "a.txt",
+                FileDigest("ab929fcd5594037960792ea0b98caf5fdaf6b60645e4ef248c28db74260f393e", 5),
+            ),
+            SymlinkEntry("b.ln", "b.txt"),
+        ]
+    )
+
+
+@pytest.mark.parametrize(
+    "create_digest, files, dirs",
+    [
+        pytest.param(
+            CreateDigest(
+                [
+                    FileContent("file.txt", b"four\n"),
+                    SymlinkEntry("symlink", "file.txt"),
+                    SymlinkEntry("a/symlink", "../file.txt"),
+                    SymlinkEntry("a/b/symlink", "../../file.txt"),
+                ]
+            ),
+            ("a/b/symlink", "a/symlink", "file.txt", "symlink"),
+            ("a", "a/b"),
+            id="simple",
+        ),
+        pytest.param(
+            CreateDigest(
+                [
+                    FileContent("file.txt", b"four\n"),
+                    SymlinkEntry("circular", "./circular"),  # After so many traversals, we give up
+                    SymlinkEntry(
+                        "a/symlink", "file.txt"
+                    ),  # looks for a/file.txt, which doesn't exist
+                    SymlinkEntry("a/too-far.ln", "../../file.txt"),  # went too far up
+                    SymlinkEntry("too-far.ln", "../file.txt"),  # went too far up
+                ]
+            ),
+            ("file.txt",),
+            ("a",),
+            id="ignored",
+        ),
+        pytest.param(
+            CreateDigest(
+                [
+                    FileContent("file.txt", b"four\n"),
+                    SymlinkEntry("a/b/parent-file.ln", "../../file.txt"),
+                    SymlinkEntry("dirlink", "a"),
+                ]
+            ),
+            ("a/b/parent-file.ln", "dirlink/b/parent-file.ln", "file.txt"),
+            ("a", "a/b", "dirlink", "dirlink/b"),
+            id="parentdir-in-symlink-target",
+        ),
+        pytest.param(
+            CreateDigest(
+                [
+                    FileContent("a/file.txt", b"four\n"),
+                    SymlinkEntry("dirlink", "a"),
+                    SymlinkEntry("double-dirlink", "dirlink"),
+                ]
+            ),
+            ("a/file.txt", "dirlink/file.txt", "double-dirlink/file.txt"),
+            ("a", "dirlink", "double-dirlink"),
+            id="double-dirlink",
+        ),
+        pytest.param(
+            CreateDigest(
+                [
+                    FileContent("file.txt", b"four\n"),
+                    SymlinkEntry("a/dirlink", "../b/c/d"),
+                    SymlinkEntry("b/c/d/link", "../../file.txt"),
+                ]
+            ),
+            # NOTE: No "b/c/d/link", because that'd result in "a/file.txt"
+            ("a/dirlink/link", "file.txt"),
+            ("a", "a/dirlink", "b", "b/c", "b/c/d"),
+            id="symlink-makes-updir-valid",
+        ),
+    ],
+)
```

Comment:
I'll be happy with A LOT more test cases

Reply:
Do any of the other REAPI-implementing build tools have test cases to "borrow"?

Reply:
Ooohhh good idea

Reply:
Although, [what an abomination of code.](https://github.com/buildbarn/bb-remote-execution/blob/71af939fd205d5411aa6d4ae89d786dc9363cd84/pkg/builder/output_hierarchy_test.go#L257) It's more abstract art than code.